server/auth/oidc.go

```diff
@@ -0,0 +1,175 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/coreos/go-oidc/v3/oidc"
+	"github.com/pkg/errors"
+)
+
+const (
+	DefaultUserNameCalm         = "sub"
+	AllowedAudienceDefaultValue = ""
+)
+
+var (
+	ErrEmptyIssueURL         = errors.New("empty issue URL")
+	ErrEmptyAllowedAudiences = errors.New("empty allowed audiences")
+	ErrUnknownIssuer         = errors.New("unknown issuer")
+	ErrUserNameNotFound      = errors.New("username not found")
+	ErrForbiddenAudience     = errors.New("forbidden audience")
+)
+
+type OIDCOptions struct {
+	AllowedIssueURLs string `json:"allowedIssueURLs,omitempty"`
+	AllowedAudiences string `json:"allowedAudiences,omitempty"`
+	UserNameClaim    string `json:"userNameClaim,omitempty"`
+}
+
+func (op *OIDCOptions) Validate() error {
+	if op.AllowedIssueURLs == "" {
+		return ErrEmptyIssueURL
+	}
+	if op.AllowedAudiences == "" {
+		return ErrEmptyAllowedAudiences
+	}
+	return nil
+}
+
+func (op *OIDCOptions) withDefault() {
+	if op.UserNameClaim == "" {
+		op.UserNameClaim = DefaultUserNameCalm
+	}
+}
+
+type ProviderWithVerifier struct {
+	provider *oidc.Provider
+	verifier *oidc.IDTokenVerifier
+}
+
+type OIDCProvider struct {
+	userNameClaim    string
+	allowedAudiences map[string]string
+
+	providers map[string]*ProviderWithVerifier
+}
+
+func (*OIDCProvider) AcceptParamType() string {
+	return ProviderParamTypeToken
+}
+
+func (p *OIDCProvider) Authenticate(ctx context.Context, param any) (string, error) {
+	token, ok := param.(string)
+	if !ok {
+		return "", ErrUnMatchedAuthenticationParamType
+	}
+	tokenParts := strings.Split(token, ".")
+	if len(tokenParts) != 3 {
+		return "", ErrMalformedToken
+	}
+	payload, err := base64.RawURLEncoding.DecodeString(tokenParts[1])
+	if err != nil {
+		return "", err
+	}
+	unsecureJwtPayload := &struct {
+		Issuer string `json:"iss"`
+	}{}
+	if err = json.Unmarshal(payload, unsecureJwtPayload); err != nil {
+		return "", err
+	}
+	issuer := unsecureJwtPayload.Issuer
+	oidcProvider, exist := p.providers[issuer]
+	if !exist {
+		return "", ErrUnknownIssuer
+	}
+	idToken, err := oidcProvider.verifier.Verify(ctx, token)
+	if err != nil {
+		return "", err
+	}
+	rawClaims := map[string]json.RawMessage{}
+	if err = idToken.Claims(&rawClaims); err != nil {
+		return "", err
+	}
+	rawMessage, ok := rawClaims[p.userNameClaim]
+	if !ok {
+		return "", ErrUserNameNotFound
+	}
+	var userName string
+	if err = json.Unmarshal(rawMessage, &userName); err != nil {
+		return "", err
+	}
+
+	// any of the client audience in the allowed is passed
+	audienceAllowed := false
+	audienceArr := idToken.Audience
+	for _, audience := range audienceArr {
+		if _, ok := p.allowedAudiences[audience]; ok {
+			audienceAllowed = true
+		}
+	}
```

Comment:
Do you think it's necessary to check the `permissions/scope` field to verify the permissions?

Reply:
we don't support authorization yet. We can consider it when we support authorization. :)